File: mixed/f1/group2/MergeIntervals.java
// Given an array of intervals where intervals[i] = [start_i, end_i], merge all overlapping intervals,
// and return an array of the non-overlapping intervals that cover all the intervals in the input.
// Sample 1:
// Input: intervals = [[1,3],[2,6],[8,10],[15,18]]
// Output: [[1,6],[8,10],[15,18]]
// Explanation: Since intervals [1,3] and [2,6] overlap, merge them into [1,6].

import java.util.LinkedList;
class MergeInterval {
    public static void main(String...args) {
        int[][] intervals = {{1,3}, {2,6}, {8,10},{15,18}};
        int[][] mergedIntervals = mergeIntervals(intervals);
        for (int i = 0; i < mergedIntervals.length; i++) {
            System.out.print("{");
            for (int j = 0; j < mergedIntervals[i].length; j++) {
                System.out.print(mergedIntervals[i][j] + ", ");
            }
            System.out.print("},");
        }
        System.out.println();
    }
    static int[][] mergeIntervals(int[][] intervals) {
        int min = Integer.MAX_VALUE;
        int max = Integer.MIN_VALUE;

        for (int i = 0; i < intervals.length; i++) {
            min = Math.min(min, intervals[i][0]);
            max = Math.max(max, intervals[i][0]);
        }

        int[] range = new int[max-min+1];
        for (int i = 0; i < intervals.length; i++) {
            range[intervals[i][0] - min] = Math.max(intervals[i][1] - min, range[intervals[i][0] - min]);
        }

        int start = 0, end = 0;
        LinkedList<int[]> result = new LinkedList<>();
        for (int i = 0; i < range.length; i++) {
            if (range[i] == 0) {
                continue;
            }
            if (i <= end) {
                end = Math.max(range[i], end);
            } else {
                result.add(new int[] {start+min, end+min});
                start = i;
                end = range[i];
            }
        }
        result.add(new int[] {start+min, end+min});

        return result.toArray(new int[result.size()][]);
    }
}

File: bytopic/intervals/python/group2/merge_intervals.py
"""
Given an array of intervals where intervals[i] = [starti, endi], merge all overlapping intervals, and
return an array of the non-overlapping intervals that cover all the intervals in the input.

Example:
Input: intervals = [[1,3],[2,6],[8,10],[15,18]]
Output: [[1,6],[8,10],[15,18]]
Explanation: Since intervals [1,3] and [2,6] overlap, merge them into [1,6].

Tag: 49/150
Tag: 56/2927, R19/2936 (overall frequency ranking), fb R1/50
"""


def merge(intervals: list[list[int]]) -> list[list[int]]:
    intervals = sorted(intervals, key=lambda x: x[0])

    ans = []

    for interval in intervals:
        if not ans or ans[-1][1] < interval[0]:
            ans.append(interval)
        else:
            ans[-1][1] = max(ans[-1][1], interval[1])
    return ans


def main():
    intervals = [[1, 3], [2, 6], [8, 10], [15, 18]]
    print(merge(intervals))


if __name__ == "__main__":
    main()
